synapse/lib/cell.py

```diff
@@ -4053,3 +4069,258 @@ async def _runReloadFunc(self, name, func):
+        if buf is None:  # pragma: no cover
+            logger.warning(f'Missing API key {iden} from user metadata for {s_common.ehex(user)}')
+            return None
+        kdef = s_msgpack.un(buf)  # This includes the shadow key
+        return kdef
+
+    async def listUserApiKeys(self, useriden):
+        '''
+        Get all the API keys for a user.
+
+        Args:
+            useriden (str): The user iden.
+
+        Returns:
+            list: A list of kdef values.
+        '''
+        user = await self.auth.reqUser(useriden)
+        vals = []
+        prefix = s_common.uhex(user.iden) + b'apikey'
+        for lkey, valu in self.slab.scanByPref(prefix, db=self.usermetadb):
+            kdef = s_msgpack.un(valu)
+            kdef.pop('shadow')
+            vals.append(kdef)
+            await asyncio.sleep(0)
+        return vals
+
+    async def getApiKeys(self):
+        '''
+        Get all API keys in the cell.
+
+        Yields:
+            tuple: Useriden, kdef values
+        '''
+        # yield all users API keys
+        for keyiden, useriden in self.slab.scanByFull(db=self.apikeydb):
+            lkey = useriden + b'apikey' + keyiden
+            valu = self.slab.get(lkey, db=self.usermetadb)
+            kdef = s_msgpack.un(valu)
+            kdef.pop('shadow')
+            yield kdef
+            await asyncio.sleep(0)
+
+    async def checkUserApiKey(self, key):
+        '''
+        Check if a user API Key is valid.
+
+        Notes:
+            If the key is not valid, the dictionary will contain a ``mesg`` key.
+            If the key is valid, the dictionar will contain the user def in a ``udef`` key,
+            and the key metadata in a ``kdef`` key.
+
+        Args:
+            key (str): The API key to check.
+
+        Returns:
+            tuple: Tuple of two items, a boolean if the key is valid and a dictionary.
+        '''
+        isok, valu = s_passwd.parseApiKey(key)
+        if isok is False:
+            return False, {'mesg': 'API Key is malformed.'}
+
+        iden, secv = valu
+
+        kdef = await self.getUserApiKey(iden)
+        if kdef is None:
+            return False, {'mesg': f'API Key does not exist: {iden}'}
+
+        user = kdef.get('user')
+        udef = await self.getUserDef(user)
+        if udef is None: # pragma: no cover
+            return False, {'mesg': f'User does not exist for API Key: {iden}', 'user': user}
+
+        if udef.get('locked'):
+            return False, {'mesg': f'User associated with API Key is locked: {iden}',
+                           'user': user, 'name': udef.get('name')}
+
+        if ((expires := kdef.get('expires')) is not None):
+            if s_common.now() > expires:
+                return False, {'mesg': f'API Key is expired: {iden}',
+                               'user': user, 'name': udef.get('name')}
+
+        shadow = kdef.pop('shadow')
+        valid = await s_passwd.checkShadowV2(secv, shadow)  # TODO Timecache this API call
```

Comment:
Do we need a story for doing this later or is this TODO for this PR?

Reply:
We have a few stories that will be kicked out of my subsequent review for the release; this is one of them.